### ISSUE TITLE
tdlib 1.8.4 Under what conditions does UpdateMessageContent come?

### ISSUE THREAD

**Issue by wowka1319** [tdlib/td#2297]
I thought UpdateMessageContent always came if the edited message was last in the chat. But I was wrong. What do guarantees exist for that?

**Comment by levlam** [comment]
`UpdateMessageContent` is sent for all messages, received by the application since instance launch, and which weren't unloaded with `updateDeleteMessages`. For the last chat messages `updateChatLastMessage` is always sent additionally.

**Comment by wowka1319** [comment]
I observe `updateDeleteMessages` of two types: (is_permanent, not from_cache) and (not is_permanent, from_cache).
1. (is_permanent, not from_cache) - is real removing (on the server). right?
2. (not is_permanent, from_cache) - is unloading by tdlib (removing from local sqlite db) you are talking about above. right?
3. `is_permanent == not from_cache` always? I hasn't seen the different case.

**Comment by levlam** [comment]
> is_permanent == not from_cache always? I hasn't seen the different case.

No. If a message is removed from cache, then the action isn't permanent, but not vice versa.

**Comment by levlam** [comment]
1. No. Messages can become inaccessible if the user clears supergroup history for self or is kicked from a private supergroup. In this case both `is_permanent` and `from_cache` will be false.
2. `is_permanent == true` means that the message is deleted forever.
3. TDLib unloads unneeded messages from memory.

**Comment by wowka1319** [comment]
1. Why TDLib decides that a message is unneeded anymore to unload it?
2. `updateChatLastMessage` is sent only when the last message is replaced by another one: a) received new message or b) actual deletion of the last message and the previous one become last ? I.e. `updateChatLastMessage` is not sent if its content has changed only?

**Comment by wowka1319** [comment]
1. What does this "some time" depend on? Sometimes deletion from cache happens too early (in 2 minutes for example), before the next `updateMessageEdited`, that make `updateMessageContent` impossible. How can I increase this timeout?
2. Do I understand correctly that TDLib when unloading a message notifies the server not to send `updateMessageContent` for it anymore?

**Comment by levlam** [comment]
TDLib has "message_unload_delay" option. See https://core.telegram.org/tdlib/options.

**Comment by AYMENJD** [comment]
It is RAM.

**Comment by wowka1319** [comment]
Docs: If `updateChatLastMessage.last_message` is null, then the last message in the chat **became unknown**. Some new unknown messages might be added to the chat in this case. 
What causes a message to become unknown?

**Comment by levlam** [comment]
The last message can be deleted, or some other unknown messages can be added to the chat.

**Comment by wowka1319** [comment]
I am seeing the following case. A new message has been posted. Since then it has been edited 5 times. I got 5 `UpdateMessageContent` and 5 `UpdateMessageEdited` (last events ~1000 seconds after posting). But I only got one `UpdateChatLastMessage` (1 second after posting). The message is still the last one. What could be causing this?

**Comment by levlam** [comment]
Send TDLib log to https://t.me/tdlib_bot.
